Client/Frontend/Browser/LightweightThemeManager.swift

```diff
@@ -0,0 +1,121 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import UIKit
+import SnapKit
+import Shared
+
+private var imgViews = WeakList<UIImageView>()
+private var instance: LightweightThemeManager? = nil
+
+@objc
+public class LightweightThemeManager {
+    private var profile: Profile
+
+    init(profile: Profile) {
+        self.profile = profile
+
+        NSNotificationCenter.defaultCenter().addObserver(self, selector: "SELThemeImageChanged:", name: "ThemeImageChanged", object: nil)
+    }
+
+    deinit {
+        NSNotificationCenter.defaultCenter().removeObserver(self, name: "ThemeImageChanged", object: nil)
+    }
+
+    private func getThemeLightness(img: UIImage?) {
+        if let img = img {
+            // Get a list of colors near the top of the image
+            let width = Int(img.size.width)
+            let height = min(10, Int(img.size.height))
+            let colors = img.getPixelData(atX: 0, andY: 0, count: width * height)
+
+            // Average the colors. We don't care about a "dominant" color here, just whether the average is light or dark.
+            var avg = [CGFloat](count: 4, repeatedValue: 0)
+            var rgb = [CGFloat](count: 4, repeatedValue: 0)
+            for color in colors {
+                color.getRed(&rgb[0], green: &rgb[1], blue: &rgb[2], alpha: &rgb[3])
+                avg[0] += rgb[0]
+                avg[1] += rgb[1]
+                avg[2] += rgb[2]
+                avg[3] += rgb[3]
+            }
+            avg[0] /= CGFloat(colors.count)
+            avg[1] /= CGFloat(colors.count)
+            avg[2] /= CGFloat(colors.count)
+            avg[3] /= CGFloat(colors.count)
+
+            // Now convert to hsl and use lightness to determine if this is light or not.
+            var c = UIColor(red: avg[0], green: avg[1], blue: avg[2], alpha: avg[3])
+            var hsl = [CGFloat](count: 4, repeatedValue: 0)
+            c.getHue(&hsl[0], saturation: &hsl[1], brightness: &hsl[2], alpha: &hsl[3])
+
+            self.profile.prefs.setBool(hsl[2] > 0.75, forKey: "themeImageIsLight")
+        } else {
+            self.profile.prefs.setBool(true, forKey: "themeImageIsLight")
+        }
+    }
+
+    func SELThemeImageChanged(notification: NSNotification) {
+        let imgName = profile.backgroundImage
+
+        let img = UIImage(named: imgName)
+        if img != nil {
+        }
+
+        for view in imgViews {
+            view.image = img
+        }
+    }
+
+    static var themeImageIsLight: Bool {
+        return instance?.profile.prefs.boolForKey("themeImageIsLight") ?? false
+    }
+
+    class func setupInstance(profile: Profile) {
+        if instance == nil {
+            instance = LightweightThemeManager(profile: profile)
+        }
+    }
+
+    public class func addThemeBackgroundTo(view: UIView, viewController: UIViewController?) -> UIView {
+        let imgView = UIImageView()
```

Comment:
I really just wanted to reduce complexity and avoid a lot of duplicated code here. i.e. only a single line of the 16 in addThemeBackgroundTo() involves setting the actual image. Likewise, when the theme image changes, we'd have to have every ViewController implement a duplicate applyTheme method. i.e. keeping them all in sync is a bit rough. I'm going to try just making a separate ThemeImageView: UIView class and see if that can absorb most of this...